detox/test/e2e/17.datePicker.test.js

```diff
@@ -1,27 +1,49 @@
-describe(':ios: DatePicker', () => {
+const jestExpect = require('expect').default;
+
+describe('DatePicker', () => {
+  beforeEach(async () => {
+    await device.reloadReactNative();
+    await element(by.text('DatePicker')).tap();
+  });
+
+  describe.each([
+    ['ios', 'compact', 0],
+    ['ios', 'inline', 1],
+    ['ios', 'spinner', 2],
+    ['android', 'calendar', 0],
+    ['android', 'spinner', 1],
+  ])(`:%s: %s mode`, (platform, mode, times) => {
     beforeEach(async () => {
-      await device.reloadReactNative();
-      await element(by.text('DatePicker')).tap();
+      for (let i = 0; i < times; i++) {
+        await element(by.id('toggleDatePicker')).tap();
+      }
     });
 
-    it('setColumnToValue should not work for a date picker', async () => {
-      let failed = false;
-      try {
-        await element(by.id('datePicker')).setColumnToValue(1, "6");
-        await element(by.id('datePicker')).setColumnToValue(2, "34");
-        await expect(element(by.id('localTimeLabel'))).toHaveText('Time: 06:34');
-      } catch(ex) {
-        failed = true;
-      }
+    test.each([
+      ['ISO8601', new Date(2019, 1, 6, 14, 10, 0).toISOString()],
+      ['yyyy/MM/dd HH:mm', '2019/02/06 14:10'],
+    ])('can select dates in %j format', async (formatString, dateString) => {
+      if (platform === 'ios') {
+        await element(by.id('datePicker')).setDatePickerDate(dateString, formatString);
+        await expect(element(by.id('localDateLabel'))).toHaveText('Date (Local): Feb 6th, 2019');
+        await expect(element(by.id('localTimeLabel'))).toHaveText('Time (Local): 2:10 PM');
+      } else {
+        await element(by.id('openDatePicker')).tap();
+        //rn-datepicker does not support testId's on android, so by.type is the only way to match the datepicker right now
+        //@see https://github.com/react-native-datetimepicker/datetimepicker#view-props-optional-ios-only
+        await element(by.type('android.widget.DatePicker')).setDatePickerDate('2019-02-06T05:10:00-08:00', 'ISO8601');
+        await element(by.text('OK')).tap();
 
-      if(failed === false) {
-        throw new Error('Test should have thrown an error, but did not');
+        await waitFor(element(by.id('utcDateLabel'))).toHaveText('Date (UTC): Feb 6th, 2019').withTimeout(3000);
```

Comment:
@noomorph 
This does not test the custom pattern feature on android, does it?

Also what I learned during my local testing, when I ran the tests between 00:00 and 01:00 is that I think you have to check against:
 - utc for iso datestrings or  custom patterns that contain explicit timezones
 - local time for patterns that don't contain patterns.
 Otherwise tests can fail due to timezone errors.

The reason is when:
- using a iso string, you let the device parse the date with an explicit format. When you then display it in the local format and the local format differs from utc, you can have a day off error around midnight. When you hoever display it in utc, the results will be stable.
- using a pattern without timezone you parse it in local time. When you now display it in utc, the result can vary when the device has a non utc timezone. So in this case the result will only be stable when you also output it as local time.

As this is quite complex, I think it is easier to remove the loop, even though the resulting code duplication is not so nice.

This was my suggestion, you may have not seen it because we worked on similar times on this:
```
      it(':ios: can select dates on a UIDatePicker, format: ISO8601', async () => {
        await element(by.id('datePicker')).setDatePickerDate('2019-02-06T05:10:00-08:00', 'ISO8601');
        await expect(element(by.id('utcDateLabel'))).toHaveText('Date (UTC): Feb 6th, 2019');
        await expect(element(by.id('utcTimeLabel'))).toHaveText('Time (UTC): 1:10 PM');
      });

      it(':ios: can select dates on a UIDatePicker, format: yyyy/MM/dd HH:mm', async () => {
        await element(by.id('datePicker')).setDatePickerDate('2019/02/06 13:10', 'yyyy/MM/dd HH:mm');
        await expect(element(by.id('localDateLabel'))).toHaveText('Date (Local): Feb 6th, 2019');
        await expect(element(by.id('localTimeLabel'))).toHaveText('Time (Local): 1:10 PM');
      });
    
      it(':android: can select dates on a UIDatePicker, format: ISO8601', async () => {
        //rn-datepicker does not support testId's on android, so by.type is the only way to match the datepicker right now
        //@see https://github.com/react-native-datetimepicker/datetimepicker#view-props-optional-ios-only
        await element(by.type('android.widget.DatePicker')).setDatePickerDate('2019-02-06T05:10:00-08:00', 'ISO8601');
        await element(by.text('OK')).tap();

        await waitFor(element(by.id('utcDateLabel'))).toHaveText('Date (UTC): Feb 6th, 2019').withTimeout(3000);
      });

      it(':android: can select dates on a UIDatePicker, format: yyyy/MM/dd', async () => {
        await element(by.type('android.widget.DatePicker')).setDatePickerDate('2019/02/06', 'yyyy/MM/dd');
        await element(by.text('OK')).tap();

        await waitFor(element(by.id('localDateLabel'))).toHaveText('Date (Local): Feb 6th, 2019').withTimeout(3000);
      });
```